stores/redis/redis.go

```diff
@@ -0,0 +1,344 @@
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/mxschmitt/golang-url-shortener/stores/shared"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	entryPathPrefix     = string("entry:")       // prefix for path-to-url mappings
+	entryUserPrefix     = string("user:")        // prefix for path-to-user mappings
+	userToEntriesPrefix = string("userEntries:") // prefix for user-to-[]entries mappings (redis SET)
+	entryVisitsPrefix   = string("entryVisits:") // prefix for entry-to-[]visit mappings (redis LIST)
+)
+
+// RedisStore implements the stores.Storage interface
+type RedisStore struct {
+	c *redis.Client
+}
+
+func New(hostaddr, password string) (*RedisStore, error) {
+	c := redis.NewClient(&redis.Options{
+		Addr:     hostaddr,
+		Password: password,
+		DB:       0,
+	})
+	// if we can't talk to redis, fail fast
+	_, err := c.Ping().Result()
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not connect to redis db0")
+	}
+	ret := &RedisStore{c: c}
+	return ret, nil
+}
+
+func (r *RedisStore) keyExists(key string) (exists bool, err error) {
+	var result *redis.IntCmd
+	logrus.Debugf("Checking for existence of key: %s", key)
+	result = r.c.Exists(key)
+	if result.Err() != nil {
+		msg := fmt.Sprintf("Error looking up key '%s': '%v', got val: '%d'", key, result.Err(), result.Val())
+		logrus.Error(msg)
+		return false, errors.Wrap(result.Err(), msg)
+	}
+	if result.Val() == 1 {
+		logrus.Debugf("Key '%s' exists!", key)
+		return true, nil
+	}
+	logrus.Debugf("Key '%s' does not exist!", key)
+	return false, nil
+}
+
+func (r *RedisStore) setValue(key string, raw []byte) error {
+	var status *redis.StatusCmd
+	logrus.Debugf("Setting value for key '%s: '%s''", key, raw)
+	status = r.c.Set(key, raw, 0) // n.b. expiration 0 means never expire
+	if status.Err() != nil {
+		msg := fmt.Sprintf("Got an unexpected error adding key '%s': %s", key, status.Err())
+		logrus.Error(msg)
+		return errors.Wrap(status.Err(), msg)
+	}
+	return nil
+}
+
+func (r *RedisStore) createValue(key string, raw []byte) error {
+	// a wrapper around setValue that returns an error if the key already exists
+	logrus.Debugf("Creating key '%s'", key)
+	exists, err := r.keyExists(key)
+	if err != nil {
+		msg := fmt.Sprintf("Could not check existence of key '%s': %s", key, err)
+		logrus.Error(msg)
+		return errors.Wrap(err, msg)
+	}
+	if exists == true {
+		msg := fmt.Sprintf("Could not create key '%s':  already exists", key)
+		logrus.Error(msg)
+		return errors.New(msg)
+	}
+	return r.setValue(key, raw)
+}
+
+func (r *RedisStore) delValue(key string) error {
+	var status *redis.IntCmd
+	logrus.Debugf("Deleting key '%s'", key)
+
+	exists, err := r.keyExists(key)
+	if err != nil {
+		msg := fmt.Sprintf("Could not check existence of key '%s': %s", key, err)
+		logrus.Error(msg)
+		return errors.Wrap(err, msg)
+	}
+	if exists == false {
+		logrus.Warnf("Tried to delete key '%s' but it's already gone", key)
+		return err
+	}
+
+	status = r.c.Del(key)
```

Comment:
I would simplify the Variable declarations to `:=` and remove the declaration 2 lines above.

Reply:
done